src/frontend/js/classes/sidebar-filter.js

```diff
@@ -0,0 +1,341 @@
+    this.sidebarContent = null;
+    this.search = null;
+    this.searchResults = [];
+    this.selectedSearchResultIndex = null;
+  }
+
+  /**
+   * Initialize sidebar filter.
+   *
+   * @param {HTMLElement[]} sections - Array of sections.
+   * @param {HTMLElement} sidebarContent - Sidebar content.
+   * @param {HTMLElement} search - Search input.
+   * @param {Function} setSectionCollapsed - Function to set section collapsed.
+   */
+  init(sections, sidebarContent, search, setSectionCollapsed) {
+    // Store refs to HTML elements.
+    this.sections = sections;
+    this.sidebarContent = sidebarContent;
+    this.search = search;
+    this.setSectionCollapsed = setSectionCollapsed;
+    let className = SidebarFilter.CSS.sidebarSearchWrapperOther;
+
+    // Search initialize with platform specific shortcut.
+    if (window.navigator.userAgent.indexOf('Mac') != -1) {
+      className = SidebarFilter.CSS.sidebarSearchWrapperMac;
+    }
+    this.search.parentElement.classList.add(className);
+
+    // Initialize search input.
+    this.search.value = '';
+    // Initialize the search results.
+    this.filter('');
+
+    // Add event listener for search input.
+    this.search.addEventListener('input', e => {
+      e.stopImmediatePropagation();
+      e.preventDefault();
+      this.filter(e.target.value);
+    });
+    // Add event listener for keyboard events.
+    this.search.addEventListener('keydown', e => this.handleKeyboardEvent(e));
+  }
+
+  /**
+   * Handle keyboard events while search input is focused.
+   *
+   * @param {Event} e - Event Object.
+   */
+  handleKeyboardEvent(e) {
+    // Return if search is not focused.
+    if (this.search !== document.activeElement) {
+      return;
+    }
+
+    // handle enter key when item is focused.
+    if (e.code === 'Enter' && this.selectedSearchResultIndex !== null) {
+      // navigate to focused item.
+      this.searchResults[this.selectedSearchResultIndex].element.click();
+      // prevent default action.
+      e.preventDefault();
+      e.stopPropagation();
+      e.stopImmediatePropagation();
+    }
+
+    // handle up and down navigation.
+    if (e.code === 'ArrowUp' || e.code === 'ArrowDown') {
+      // check for search results.
+      if (this.searchResults.length === 0) {
+        return;
+      }
+
+      // get current focused item.
+      const prevSelectedSearchResultIndex = this.selectedSearchResultIndex;
+
+      // get next item to be focus.
+      this.selectedSearchResultIndex = this.getNextTitleOrItemIndex(e.code,
+        this.selectedSearchResultIndex,
+        this.searchResults.length - 1);
+
+      // blur previous focused item.
+      this.blurTitleOrItem(prevSelectedSearchResultIndex);
+      // focus next item.
+      this.focusTitleOrItem(this.selectedSearchResultIndex);
+
+      // prevent default action.
+      e.preventDefault();
+      e.stopPropagation();
+      e.stopImmediatePropagation();
+    }
+  }
+
+  /**
+   * Get next title or item index.
```

Comment:
I don't understand when (and why) it can return either `title` or `index`? Why not just index?

Reply:
just confusing docs changed